### ISSUE TITLE
NBTTagCompound

### ISSUE THREAD

**Issue by xxlio109** [SpongePowered/SpongeForge#2083]
SpongeForge version: spongeforge-1.12.2-2611-7.1.0-BETA-3002

Forge version: forge-1.12.2-14.23.2.2611-universal

Java version: openjdk version "1.8.0_161"

Operating System: centos 7

Plugins/Mods:

Minecraft, Minecraft Coder Pack, Forge Mod Loader, Minecraft Forge, SpongeAPI, SpongeForge, Actually Additions, Advanced Generators, BD Lib, Baubles, Binnie Core, Binnie's Botany, Binnie's Design, Binnie's Extra Bees, Binnie's Extra Trees, Binnie's Genetics, Blood Magic: Alchemical Wizardry, Botania, Brandon's Core, CoFH Core, CoFH World, CodeChicken Lib, Cyclops Core, Draconic Evolution, Erebus, EvilCraft, EvilCraft-Compat, FTB Utilities, FTBLib, FoamFix, FoamFixCore, Forestry, Galacticraft Core, Galacticraft Planets, Guide-API, Immersive Engineering, Improved Extraction, Industrial Foregoing, IndustrialCraft 2, Iron Chest, Just Enough Items, MJRLegendsLib, Mantle, MeanMobs, Mekanism, MekanismGenerators, MekanismTools, Micdoodle8 Core, More Planets, NetherPortalFix, NuclearCraft, OMLib, Open Modular Passive Defense, Open Modular Turrets, Planet Progression, PlusTiC, Reborn Core, Redstone Flux, Reforged, Shadowfacts' Forgelin, TESLA, Tech Reborn, Tesla Core Lib, Tesla Core Lib Registries, The Betweenlands, Thermal Cultivation, Thermal Dynamics, Thermal Expansion, Thermal Foundation, Tinker I/O, Tinkers' Construct, UniDict, WanionLib

Issue Description
NBTTagCompound

Error log:
https://gist.github.com/xxlio109/6d40ee56b37f85b5e2f41aa0715006aa

**Comment by phit** [comment]
looks like a mod issue to me, please report it to Galacticraft 

```
Details:
	Entity Type: null (micdoodle8.mods.galacticraft.core.entities.player.GCEntityPlayerMP)
	Entity ID: 6433
	Entity Name: xie_yue
```

**Comment by radfast** [comment]
Hi, Galacticraft developer here.

This is **not** a Galacticraft-related crash, on the information available in that crash report.

@phit : the suggestion that it is Galacticraft, is disturbing.  Although I can see that the crash report contains this line for the entity being ticked:

    Entity Type: null (micdoodle8.mods.galacticraft.core.entities.player.GCEntityPlayerMP)

that is merely stating the entity being ticked at the time, i.e. the player.  In any crash report, the player entity will **always** be `GCEntityPlayerMP` if Galacticraft is installed.  So all that the crash report is really telling you is that the error occurred while ticking the player.

I repeat, that does not mean it is our code crashing or that there is any link with our code at all.  [GCEntityPlayerMP](https://github.com/micdoodle8/Galacticraft/blob/MC1.12/src/main/java/micdoodle8/mods/galacticraft/core/entities/player/GCEntityPlayerMP.java) passes through almost all methods to the superclass `EntityPlayerMP`, so something like `onLivingUpdate()` will pass straight through our code into the vanilla code.  Seems to be vanilla code or Sponge crashing here.

(Our player subclass `GCEntityPlayerMP` is used as a convenient way to hook into several `EntityPlayerMP` methods that we need to hook into.  It has been said by other devs that this risks compatibility issues with other mods, because if all mods did this then things might break, and instead we should hook separately into each method.  But for now it works very well for us as it is, it is widely compatible because it does not interfere with any other mod's changes to `EntityPlayerMP`.  In contrast, hooking separately into each method in `EntityPlayerMP` could create issues if other mods are also making ASM changes to those same methods.  The "proper" solution is for all mods to use **PlayerAPI**.  We offer full PlayerAPI compatibility if it is present.  But I've been told that Sponge and PlayerAPI are not compatible with each other ....

Frankly the one reason we might change this in future is exactly to prevent issues like this, where something else is crashing while ticking `EntityPlayerMP` and everybody "thinks" it is Galacticraft instability because the crash report contains `GCEntityPlayerMP`.)

**Comment by radfast** [comment]
TL;DR:  You need to re-open this.  It is not a Galacticraft issue.

**Comment by phit** [comment]
my bad, I shouldn't look at issues when I just woke up.. still don't see this pointing to Sponge though

@xxlio109 please ask xie_yue what they were doing when this occured

**Comment by radfast** [comment]
PS  one thing I noticed, the instance which crashed has both **SpongeForge** (normally a server-side mod) and **JEI** and **FoamFix** (normally client-side mods).  Seems a bit strange to have all these together.


Here is [another similar crash report](https://pastebin.com/98fyuJFf) I found.  This one also has Sponge, JEI and Foamfix all together, and some overlapping mods.

**Comment by gabizou** [comment]
It sounds like an NBTTagCompound contains itself which is causing recursion. I'd vaguely try to figure that sponge isn't doing much (Sponge added a null check into NBTTagCompound because there were some mods that were storing nulls into NBTTagCompound and when the `#copy()` method was called, it'd NPE).

**Comment by gabizou** [comment]
Ok, now that it has been built, build 3004 has some improved debugging output for this specific case, if you can replicate the issue easily, it should print out some nice logs for us to read and understand.

**Comment by xxlio109** [comment]
New crash ，I don't know if it's the same as before。。。
spongeforge-1.12.2-2611-7.1.0-BETA-3004
https://gist.github.com/xxlio109/2bd19884782dfb253b393f2db3cefcdb

**Comment by broeredaan** [comment]
My server crashes on startup because of this
https://gist.github.com/broeredaan/1339bee2dc66bf18246766faa2de7129
(edit: currently running the SkyBlock 3 V3.3 pack with spongeforge 3006)

**Comment by gabizou** [comment]
@xxlio109 different crash, that's sponge catching the error and making it more available as to what nbttag compound is being used to store a null value (which is not allowed, even by Mojang's standards).

@broeredaan that's a villager crash, related to #2048

**Comment by gabizou** [comment]
@radfast any idea where the crashlog you got came from or how to reproduce? I added stack overflow error handling for nbttagcompound writing, but it just seems like someone is adding a compound to a list to itself, but so far no one has reproduced it with my changes.

**Comment by radfast** [comment]
hey @gabizou  no idea, sorry - the pastebin came up on a Google search for some of the text seen in xxlio109's error log.  I cannot find any page linking to it or discussing the pastebin.

Because I only found it in that one search - and here - this suggests to me that the error is **very rare**.  Maybe wait a month and see if anything comes up in your logging?

**Comment by broeredaan** [comment]
@gabizou This was on server startup where the server doesnt even have a world yet. (it crashes before it can create a world because of this). But ill keep track of #2048 I guess then. Thanks

**Comment by broeredaan** [comment]
So @gabizou #2048 was closed with 3014 but I am still experiencing the issue with NBTTagCompound from the gist I posted here. So perhaps it does have something to do with this issue? 
https://gist.github.com/broeredaan/1139b6c2d1801cc2e1f3f3ae02db690d
Or should I make a new issue for this?

**Comment by gabizou** [comment]
I'm going to say it once, #2048 is a *separate bug/issue that was resolved with separate code*.

[This line is saying it's a mod's fault](https://gist.github.com/broeredaan/1139b6c2d1801cc2e1f3f3ae02db690d#file-another-L10-L14) for adding a null **as the log suggests**, I wrote the message to catch nulls for mods to stop breaking things, which [buildcraft had an issue at some point that they couldn't figure out how until we reported it with the null check](https://github.com/BuildCraft/BuildCraft/issues/3213).

In this particular case, we can't pinpoint what is going on, so I may add some additional cause stack information, but that may spam the server (@phit thinking of adding some additional things like phase tracker stack dump and whatnot).

@broeredaan but in all honesty, I didn't do anything to **fix** this issue, I simply added some additional logging in place where it is failing, but what you're reporting now is closer related to the build craft issue where a mod Is adding a `null`, this issue is centered around something causing recursive nbt loops causing the server to crash.
